### PR TITLE
Loki: Visually distinguish error logs for LogQL2

### PR DESCRIPTION
**Pull request by ivanahuckova**
**What this PR does / why we need it**:
In LogQL2, when querying with a given pipeline such as `{app="foo"} | logfmt | duration > 240ms` multiple kind of errors for each line could happens, for instance if the line is not in logfmt or if the value duration label is not a duration. In Loki we don't filter out those log line, this would be confusing and dangerous to filter line that failed to be processed. Instead we add a label `__ error __`, however with the current UI it's not clear this is happening. 

To inform users about this, we are making those lines more pale and on tooltip you can see the content of `__error__`. That way it's clear that they are not being considered. We are also adding error message on top informing users about the fact, that some of the logs were errored. 

image

### CODE REVIEW

File: public/app/plugins/datasource/loki/result_transformer.ts [hunk 1]
```diff
@@ -325,6 +325,10 @@ export function lokiStreamsToDataframes(
     const dataFrame = lokiStreamResultToDataFrame(stream, reverse);
     enhanceDataFrame(dataFrame, config);
 
+    if (meta.custom && dataFrame.fields.some(f => f.labels && Object.keys(f.labels).some(l => l === '__error__'))) {
+      meta.custom.error = 'Error when parsing some of the logs';
```
Comment (cyriltovena) [inline public/app/plugins/datasource/loki/result_transformer.ts:329]: This is a good starting point. I think later we could add more descriptive information.

**Review by cyriltovena — APPROVED** [review]
LGTM

File: packages/grafana-data/src/types/logs.ts [hunk 2]
```diff
@@ -31,6 +31,7 @@ export enum LogsMetaKind {
   Number,
   String,
   LabelsMap,
+  Error,
```
Comment (ivanahuckova) [inline packages/grafana-data/src/types/logs.ts:34]: Added so we can have special styling for errors - e.g. red color.

File: packages/grafana-ui/src/components/Logs/LogRow.tsx [hunk 3]
```diff
@@ -157,25 +182,29 @@ class UnThemedLogRow extends PureComponent<Props, State> {
...
 
     return (
       <>
-        <tr
-          className={hoverBackground}
-          onMouseEnter={this.addHoverBackground}
-          onMouseLeave={this.clearHoverBackground}
-          onClick={this.toggleDetails}
-        >
+        <tr className={logRowBackground} onClick={this.toggleDetails}>
```
Comment (ivanahuckova) [inline packages/grafana-ui/src/components/Logs/LogRow.tsx:195]: I have removed `onMouseEnter` and `onMouseLeave` as it was breaking the tooltip - tooltip often wasn't showing. We are still passing these methods to Log Details so if we move with cursor to log details, log row is going to get the background color as well.

File: packages/grafana-ui/src/components/Logs/LogRow.tsx [hunk 4]
```diff
@@ -72,6 +76,10 @@ const getStyles = stylesFactory((theme: GrafanaTheme) => {
       label: hoverBackground;
       background-color: ${bgColor};
     `,
+    erroredLog: css`
```
Comment (davkal) [inline packages/grafana-ui/src/components/Logs/LogRow.tsx:79]: Is this a typo?

File: public/app/features/explore/Logs.tsx [hunk 5]
```diff
@@ -40,6 +40,8 @@ function renderMetaItem(value: any, kind: LogsMetaKind) {
         <LogLabels labels={value} />
       </span>
     );
+  } else if (kind === LogsMetaKind.Error) {
+    return <span className="logs-meta-item__error">{value}</span>;
```
Comment (davkal) [inline public/app/features/explore/Logs.tsx:44]: I wonder if a full width error message would not be better. But we can improve this in a later PR.

**Review by davkal — COMMENTED** [review]
Works well and looks good. Minor nits/questions though.

File: packages/grafana-data/src/types/logs.ts (hunk 2 above)
Comment (aocenas) [reply packages/grafana-data/src/types/logs.ts:34]: Can we add some comment what does LogsMetaKind mean? I was a bit confused whether it is some log type or what's the difference from log level.

**Comment by aocenas** [comment]
Just to understand the `__error__` label is something that Loki adds in the response, right?

**Comment by ivanahuckova** [comment]
>Just to understand the __error__ label is something that Loki adds in the response, right?

It will add   __error__ label to every log that wasn't able to be parsed.

File: packages/grafana-data/src/types/logs.ts [hunk 6]
```diff
@@ -27,10 +27,12 @@ export enum LogLevel {
   unknown = 'unknown',
 }
 
+// Used for meta information in logs view in Explore
```
Comment (ivanahuckova) [inline packages/grafana-data/src/types/logs.ts:30]:
```suggestion
// Used for meta information such as common labels or returned log rows in logs view in Explore
```

File: public/app/features/explore/Logs.tsx (hunk 5 above)
Comment (ivanahuckova) [reply public/app/features/explore/Logs.tsx:44]: 👍

File: packages/grafana-ui/src/components/Logs/LogRow.tsx [hunk 7]
```diff
@@ -86,8 +94,14 @@ class UnThemedLogRow extends PureComponent<Props, State> {
     showContext: false,
     showDetails: false,
     hasHoverBackground: false,
+    hasError: false,
+    errorTooltip: undefined,
   };
 
+  componentDidMount() {
+    this.setErrorAndTooltip();
```
[the remark this answers is not in the document]
Comment (ivanahuckova) [reply packages/grafana-ui/src/components/Logs/LogRow.tsx]: Yes, you are right. I overcomplicated it. See refactor and simplification: https://github.com/grafana/grafana/pull/28359/commits/3bbe55d76128c6b86af53ad98ee2784b45f4809e

File: packages/grafana-ui/src/components/Logs/LogDetails.tsx [hunk 8]
```diff
@@ -88,6 +90,8 @@ class UnThemedLogDetails extends PureComponent<Props> {
     const labelsAvailable = Object.keys(labels).length > 0;
     const fields = getAllFields(row, getFieldLinks);
     const parsedFieldsAvailable = fields && fields.length > 0;
+    // If logs with error, we are not showing the level color
+    const levelClassName = !hasError ? cx(style.logsRowLevel, styles.logsRowLevelDetails) : '';
```
Comment (Elfo404) [inline packages/grafana-ui/src/components/Logs/LogDetails.tsx:94]:
```suggestion
    const levelClassName = cx(!hasError && [style.logsRowLevel, styles.logsRowLevelDetails]);
```

**Review by Elfo404 — APPROVED** [review]
LGTM, just a tiny nit :)

**Pull request merged**